client.go

```diff
@@ -3,40 +3,47 @@ package dns
 // A client implementation.
 
 import (
-	"bytes"
 	"context"
 	"crypto/tls"
 	"encoding/binary"
+	"fmt"
 	"io"
 	"net"
+	"strings"
 	"time"
 )
 
-const dnsTimeout time.Duration = 2 * time.Second
-const tcpIdleTimeout time.Duration = 8 * time.Second
+const (
+	dnsTimeout     time.Duration = 2 * time.Second
+	tcpIdleTimeout time.Duration = 8 * time.Second
+)
 
 // A Conn represents a connection to a DNS server.
 type Conn struct {
 	net.Conn                         // a net.Conn holding the connection
 	UDPSize        uint16            // minimum receive buffer for UDP messages
-	TsigSecret     map[string]string // secret(s) for Tsig map[<zonename>]<base64 secret>, zonename must be fully qualified
-	rtt            time.Duration
-	t              time.Time
+	TsigSecret     map[string]string // secret(s) for Tsig map[<zonename>]<base64 secret>, zonename must be in canonical form (lowercase, fqdn, see RFC 4034 Section 6.2)
+	TsigProvider   TsigProvider      // An implementation of the TsigProvider interface. If defined it replaces TsigSecret and is used for all TSIG operations.
 	tsigRequestMAC string
 }
 
 // A Client defines parameters for a DNS client.
 type Client struct {
-	Net            string                                                                 // if "tcp" or "tcp-tls" (DNS over TLS) a TCP query will be initiated, otherwise an UDP one (default is "" for UDP)
-	UDPSize        uint16                                                                 // minimum receive buffer for UDP messages
-	TLSConfig      *tls.Config                                                            // TLS connection configuration
-	Timeout        time.Duration                                                          // a cumulative timeout for dial, write and read, defaults to 0 (disabled) - overrides DialTimeout, ReadTimeout and WriteTimeout when non-zero
-	DialTimeout    time.Duration                                                          // net.DialTimeout, defaults to 2 seconds - overridden by Timeout when that value is non-zero
-	ReadTimeout    time.Duration                                                          // net.Conn.SetReadTimeout value for connections, defaults to 2 seconds - overridden by Timeout when that value is non-zero
-	WriteTimeout   time.Duration                                                          // net.Conn.SetWriteTimeout value for connections, defaults to 2 seconds - overridden by Timeout when that value is non-zero
-	TsigSecret     map[string]string                                                      // secret(s) for Tsig map[<zonename>]<base64 secret>, zonename must be fully qualified
-	SingleInflight bool                                                                   // if true suppress multiple outstanding queries for the same Qname, Qtype and Qclass
-	Dial           func(network, address string, timeout time.Duration) (net.Conn, error) // optional function for dialing
```

Comment:
Our `Dial` had to be renamed to `CustomDial` since `Dial` is now a function on the client.